File: basics/02-linkedlist/02-86.py
"""
check the singly linked list

must break the original list, or there wil be circle
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        before, after = ListNode(0), ListNode(0)
        before_cur, after_cur = before, after
        while head:
            if head.val < x:
                before_cur.next = head
                before_cur = before_cur.next
            else:
                after_cur.next = head
                after_cur = after_cur.next
            #
            tmp = head.next
            head.next = None
            head = tmp

        before_cur.next = after.next
        return before.next

File: basics/02-linkedlist/02-86.cpp
/*
 * leetcode 86
 *
 * must break original linked list
 *
 */

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* dummy1 = new ListNode(-1);
        ListNode* dummy2 = new ListNode(-1);
        ListNode* p = head;
        ListNode* p1 = dummy1;
        ListNode* p2 = dummy2;
        while(p != nullptr) {
            if (p->val < x) {
                p1->next = p;
                p1 = p1->next;
            } else {
                p2->next = p;
                p2 = p2->next;
            }
            // break p
            ListNode* t = p->next;
            p->next = nullptr;
            p = t;
        }
        p1->next = dummy2->next;
        return dummy1->next;    
    }
};

